### ISSUE TITLE
Carlini L2 and L-inf are slow

### ISSUE THREAD

**Issue by siarez** [Trusted-AI/adversarial-robustness-toolbox#1060]
This is a question:
I have manually implemented my own Carlini method. I'm double checking my work by using this libarary. (Great work btw!) 
I find it that it takes a very long time for the `generate` method to return. The L2 method specifically takes 8.5 minutes.
I'm only attacking 256 images, on a RTX2080ti, and GPU utilization is at 98%.
My hyper parameters are:
`batch_size=256, 
targeted=True, 
learning_rate=0.00005, 
max_iter=6,  
initial_const=0.1`

I have also tried 
`batch_size=256, 
targeted=True, 
learning_rate=0.01, 
max_iter=6,  
initial_const=0.01`

My own implementation takes about 2 minutes.
I also don't have this issue with the targeted `FastGradientMethod`
Any insights are appreciated.

**Comment by beat-buesser** [comment]
Hi @siarez 

Thank you very much for using and testing ART and thank you very much for the compliments to the ART team!

It is definitely possible that certain tools of ART, especially the ones implemented framework agnostic, could be further optimised. 

Would you be able to share more details of your implementation/approach? Are both implementations achieving comparable adversarial accuracies and perturbations?

Would you be interested to contribute your faster version to ART?

**Comment by siarez** [comment]
Yes, I'm using PyTorch. 
I dug into the source code a bit. I'm still not too familiar with it, but I think I may have found the culprit (or one of the culprits). 
[Here](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/b0bd2ea4bbe466f3f99431a65d1e235ee405ba06/art/estimators/classification/pytorch.py#L593) it looks like the gradient is being calculated for each sample in the batch individually in a for-loop. I'm not sure why it is not calculated for the whole batch all at once. That should speed things up.
Maybe the person who made the commit (@ririnicolae) can shed some light?

**Comment by siarez** [comment]
Yes, I'm using PyTorch.
I dug into the source code a bit. I'm still not too familiar with it, but I think I may have found the culprit (or one of the culprits).
Here it looks like the gradient is being calculated for each sample in the batch individually in a for-loop. I'm not sure why it is not calculated for the whole batch all at once. 
We could compute a loss that only aggregates the outputs that we are interested in. Then call backward on the loss which will do the whole batch at once. That should speed things up.
Maybe the person who made the commit (@ririnicolae) can shed some light?

**Comment by beat-buesser** [comment]
Hi @siarez 

I agree, the current implementation is calculating more gradients than it requires. Maybe it was the best solution for early versions of `torch`.

Would you be interested to implement an improved approach based on a loss that aggregates only the required outputs as you are proposing?

**Comment by siarez** [comment]
I can give it a try after the NeurIPS deadline :)

**Comment by beat-buesser** [comment]
@siarez That would be amazing. I hope everything goes well with the NeurIPS submission! Let us know when we can help or support.

**Comment by beat-buesser** [comment]
I'll add this issue to the 1.7 project to keep track of it but without adding any time pressure, we can move to it to a later release if needed.

**Comment by beat-buesser** [comment]
Hi @siarez Did you have a chance to take a look at this issue?

**Comment by siarez** [comment]
Hi @beat-buesser there were too many things that I wasn't quite getting about the code.

**Comment by beat-buesser** [comment]
@siarez Great! Do you think it would help if we discuss parts of the code to close the remaining gaps?
